File: enigma_decoder/enigma.py
"""
エニグマM3シミュレータ。

ローター I-V、リフレクター B/C をサポート。
ダブルステッピングのアノマリも正しく実装している。
"""

# 各ローターの (配線, ノッチ位置)
ROTORS = {
    'I':   ('EKMFLGDQVZNTOWYHXUSPAIBRCJ', 'Q'),
    'II':  ('AJDKSIRUXBLHWTMCQGZNPYFVOE', 'E'),
    'III': ('BDFHJLCPRTXVZNYEIWGAKMUSQO', 'V'),
    'IV':  ('ESOVPZJAYQUIRHXLNFTGKDCMWB', 'J'),
    'V':   ('VZBRGITYUPSDNHLXAWMJQOFECK', 'Z'),
}

REFLECTORS = {
    'B': 'YRUHQSLDPXNGOKMIEBFZCWVJAT',
    'C': 'FVPJIAOYEDRZXWGCTKUQSBNMHL',
}

ROTOR_NAMES = ['I', 'II', 'III', 'IV', 'V']


class Enigma:
    """3ローターのエニグマM3。"""

    def __init__(self, rotor_names=('I', 'II', 'III'),
                 reflector_name='B',
                 ring_settings=(0, 0, 0),
                 positions=(0, 0, 0),
                 plugboard=''):
        # 順方向と逆方向の配線をint配列で持つ（高速化のため）
        self.fwd = []
        self.bwd = []
        for n in rotor_names:
            wiring = [ord(c) - 65 for c in ROTORS[n][0]]
            inv = [0] * 26
            for i, v in enumerate(wiring):
                inv[v] = i
            self.fwd.append(wiring)
            self.bwd.append(inv)
        self.notches = [ord(ROTORS[n][1]) - 65 for n in rotor_names]
        self.reflector = [ord(c) - 65 for c in REFLECTORS[reflector_name]]
        self.rings = list(ring_settings)
        self.initial_positions = list(positions)
        self.positions = list(positions)
        self.plugboard = self._parse_plugboard(plugboard)

    @staticmethod
    def _parse_plugboard(s):
        pb = list(range(26))
        if isinstance(s, str):
            pairs = s.split()
        else:
            pairs = s
        for pair in pairs:
            if len(pair) == 2:
                a = ord(pair[0].upper()) - 65
                b = ord(pair[1].upper()) - 65
                pb[a], pb[b] = b, a
        return pb

    def reset(self, positions=None):
        """ローター位置をリセット。"""
        if positions is None:
            self.positions = list(self.initial_positions)
        else:
            self.positions = list(positions)

    def _step(self):
        """ダブルステッピングを含むローター進行。"""
        # 中ローターがノッチ位置にあるなら、左と中の両方が進む（ダブルステップ）
        if self.positions[1] == self.notches[1]:
            self.positions[0] = (self.positions[0] + 1) % 26
            self.positions[1] = (self.positions[1] + 1) % 26
        elif self.positions[2] == self.notches[2]:
            self.positions[1] = (self.positions[1] + 1) % 26
        # 右ローターは常に進む
        self.positions[2] = (self.positions[2] + 1) % 26

    def encrypt_int(self, c):
        """単一文字（0-25）を暗号化/復号。"""
        self._step()
        c = self.plugboard[c]
        # 右→左
        for i in (2, 1, 0):
            offset = (self.positions[i] - self.rings[i]) % 26
            c = (c + offset) % 26
            c = self.fwd[i][c]
            c = (c - offset) % 26
        # リフレクター
        c = self.reflector[c]
        # 左→右
        for i in (0, 1, 2):
            offset = (self.positions[i] - self.rings[i]) % 26
            c = (c + offset) % 26
            c = self.bwd[i][c]
            c = (c - offset) % 26
        c = self.plugboard[c]
        return c

    def encrypt(self, text):
        """文字列を暗号化/復号（A-Zのみ処理）。"""
        text = text.upper()
        out = []
        for ch in text:
            if 'A' <= ch <= 'Z':
                out.append(chr(self.encrypt_int(ord(ch) - 65) + 65))
        return ''.join(out)


def precompute_rotor_arrays(rotor_names, reflector):
    """ローター・リフレクターの配線をint配列で事前計算。"""
    fwd = [[ord(c) - 65 for c in ROTORS[n][0]] for n in rotor_names]
    bwd = []
    for w in fwd:
        inv = [0] * 26
        for i, v in enumerate(w):
            inv[v] = i
        bwd.append(inv)
    notches = [ord(ROTORS[n][1]) - 65 for n in rotor_names]
    reflector_arr = [ord(c) - 65 for c in REFLECTORS[reflector]]
    return fwd, bwd, notches, reflector_arr


def decrypt_fast(ciphertext_ints, fwd, bwd, notches, reflector_arr,
                 ring_settings, positions, plugboard_array):
    """
    内部ループ用の高速版（ローター配列は事前計算済みで渡される）。
    """
    # ローカル変数化（Python の属性参照を減らすため）
    fwd0, fwd1, fwd2 = fwd[0], fwd[1], fwd[2]
    bwd0, bwd1, bwd2 = bwd[0], bwd[1], bwd[2]
    n0, n1, n2 = notches[0], notches[1], notches[2]
    r0, r1, r2 = ring_settings[0], ring_settings[1], ring_settings[2]
    p0, p1, p2 = positions[0], positions[1], positions[2]

    out = []
    out_append = out.append
    for c in ciphertext_ints:
        # ダブルステッピング
        if p1 == n1:
            p0 = (p0 + 1) % 26
            p1 = (p1 + 1) % 26
        elif p2 == n2:
            p1 = (p1 + 1) % 26
        p2 = (p2 + 1) % 26

        x = plugboard_array[c]
        # 右→左
        x = fwd2[(x + p2 - r2) % 26]
        x = (x - p2 + r2) % 26
        x = fwd1[(x + p1 - r1) % 26]
        x = (x - p1 + r1) % 26
        x = fwd0[(x + p0 - r0) % 26]
        x = (x - p0 + r0) % 26
        # リフレクター
        x = reflector_arr[x]
        # 左→右
        x = bwd0[(x + p0 - r0) % 26]
        x = (x - p0 + r0) % 26
        x = bwd1[(x + p1 - r1) % 26]
        x = (x - p1 + r1) % 26
        x = bwd2[(x + p2 - r2) % 26]
        x = (x - p2 + r2) % 26
        out_append(plugboard_array[x])
    return out


def decrypt_with_settings(ciphertext_ints, rotor_names, reflector,
                          ring_settings, positions, plugboard_array):
    """便利な互換ラッパー（毎回事前計算するので遅い。テスト用）。"""
    fwd, bwd, notches, refl = precompute_rotor_arrays(rotor_names, reflector)
    return decrypt_fast(ciphertext_ints, fwd, bwd, notches, refl,
                        ring_settings, positions, plugboard_array)


# 既知の自己テスト
if __name__ == '__main__':
    # 既知の正解: rotors=I,II,III, reflector=B, all rings 0, position 0,
    # input AAAAA → output BDZGO
    e = Enigma(('I', 'II', 'III'), 'B', (0, 0, 0), (0, 0, 0), '')
    result = e.encrypt('AAAAA')
    assert result == 'BDZGO', f'Expected BDZGO, got {result}'
    print(f'[OK] AAAAA → {result}')

    # 対合性のテスト
    e1 = Enigma(('I', 'II', 'III'), 'B', (5, 12, 7), (3, 14, 9), 'AB CD EF')
    plaintext = 'HELLOWORLDTHISISATEST'
    ct = e1.encrypt(plaintext)
    e2 = Enigma(('I', 'II', 'III'), 'B', (5, 12, 7), (3, 14, 9), 'AB CD EF')
    pt = e2.encrypt(ct)
    assert pt == plaintext, f'Round trip failed: {plaintext} → {ct} → {pt}'
    print(f'[OK] {plaintext} → {ct} → {pt}')
    print('全テスト合格')

File: web/lib/enigma.ts
/**
 * エニグマ M3 シミュレータ（TypeScript 版）。
 *
 * enigma.py の忠実な移植。ローター I-V、リフレクター B/C、
 * ダブルステッピングのアノマリも同一に実装している。
 * 既知ベクトル AAAAA → BDZGO（rotors I,II,III / reflector B / rings 0 / pos 0）で検証。
 */

// 各ローターの [配線, ノッチ文字]
export const ROTORS: Record<string, [string, string]> = {
  I: ["EKMFLGDQVZNTOWYHXUSPAIBRCJ", "Q"],
  II: ["AJDKSIRUXBLHWTMCQGZNPYFVOE", "E"],
  III: ["BDFHJLCPRTXVZNYEIWGAKMUSQO", "V"],
  IV: ["ESOVPZJAYQUIRHXLNFTGKDCMWB", "J"],
  V: ["VZBRGITYUPSDNHLXAWMJQOFECK", "Z"],
};

export const REFLECTORS: Record<string, string> = {
  B: "YRUHQSLDPXNGOKMIEBFZCWVJAT",
  C: "FVPJIAOYEDRZXWGCTKUQSBNMHL",
};

export const ROTOR_NAMES = ["I", "II", "III", "IV", "V"];

const A = 65;

export interface EnigmaConfig {
  rotors: [string, string, string];
  reflector: "B" | "C";
  rings: [number, number, number]; // 0-25
  positions: [number, number, number]; // 0-25
  plugboard: string; // "AB CD EF" 形式
}

function parsePlugboard(s: string): number[] {
  const pb = Array.from({ length: 26 }, (_, i) => i);
  for (const pair of s.toUpperCase().split(/\s+/)) {
    if (pair.length === 2) {
      const a = pair.charCodeAt(0) - A;
      const b = pair.charCodeAt(1) - A;
      if (a >= 0 && a < 26 && b >= 0 && b < 26) {
        pb[a] = b;
        pb[b] = a;
      }
    }
  }
  return pb;
}

export class Enigma {
  private fwd: number[][] = [];
  private bwd: number[][] = [];
  private notches: number[];
  private reflector: number[];
  private rings: number[];
  private positions: number[];
  private plugboard: number[];

  constructor(cfg: EnigmaConfig) {
    for (const n of cfg.rotors) {
      const wiring = ROTORS[n][0].split("").map((c) => c.charCodeAt(0) - A);
      const inv = new Array(26).fill(0);
      wiring.forEach((v, i) => (inv[v] = i));
      this.fwd.push(wiring);
      this.bwd.push(inv);
    }
    this.notches = cfg.rotors.map((n) => ROTORS[n][1].charCodeAt(0) - A);
    this.reflector = REFLECTORS[cfg.reflector].split("").map((c) => c.charCodeAt(0) - A);
    this.rings = [...cfg.rings];
    this.positions = [...cfg.positions];
    this.plugboard = parsePlugboard(cfg.plugboard);
  }

  private step(): void {
    // ダブルステッピング: 中ローターがノッチ位置なら左と中が同時に進む
    if (this.positions[1] === this.notches[1]) {
      this.positions[0] = (this.positions[0] + 1) % 26;
      this.positions[1] = (this.positions[1] + 1) % 26;
    } else if (this.positions[2] === this.notches[2]) {
      this.positions[1] = (this.positions[1] + 1) % 26;
    }
    this.positions[2] = (this.positions[2] + 1) % 26;
  }

  private encryptInt(c: number): number {
    this.step();
    let x = this.plugboard[c];
    // 右→左
    for (const i of [2, 1, 0]) {
      const offset = ((this.positions[i] - this.rings[i]) % 26 + 26) % 26;
      x = (x + offset) % 26;
      x = this.fwd[i][x];
      x = ((x - offset) % 26 + 26) % 26;
    }
    // リフレクター
    x = this.reflector[x];
    // 左→右
    for (const i of [0, 1, 2]) {
      const offset = ((this.positions[i] - this.rings[i]) % 26 + 26) % 26;
      x = (x + offset) % 26;
      x = this.bwd[i][x];
      x = ((x - offset) % 26 + 26) % 26;
    }
    return this.plugboard[x];
  }

  /** A-Z のみ処理して暗号化/復号（対合的なので同じ操作）。 */
  encrypt(text: string): string {
    const out: string[] = [];
    for (const ch of text.toUpperCase()) {
      const code = ch.charCodeAt(0);
      if (code >= A && code <= A + 25) {
        out.push(String.fromCharCode(this.encryptInt(code - A) + A));
      }
    }
    return out.join("");
  }
}

/** ワンショットで暗号化/復号する便利関数（毎回 fresh な状態で実行）。 */
export function runEnigma(cfg: EnigmaConfig, text: string): string {
  return new Enigma(cfg).encrypt(text);
}

/** 5文字ごとに空白を入れる（軍用電文の慣習）。 */
export function group5(s: string): string {
  return (s.match(/.{1,5}/g) || []).join(" ");
}
